Unclear test suite failures when using empty xdescribe blocks

I found an annoying scenario where the test suite fails to run when a spec contains an empty `xdescribe(...)` call. I often times write a skeleton of a test suite when TDDing some new code, so I wind up with a file like this:

```kotlin
object SomeNewClassSpec : Spek({
    xdescribe("some new functionality") {
    }
    xdescribe("some other new functionality") {
    }
})
```

I'll then progressively flesh the file out, implementing tests, then implementing code, running the tests as I go and watching them start to pass. However, when trying to do so, the suite fails to run and I get no feedback on the command line, but looking at the test report HTML, I see this:

```
org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 22.
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:131)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:155)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:137)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.UnsupportedOperationException
	at org.spekframework.spek2.junit.SpekTestDescriptor.removeFromHierarchy(SpekTestDescriptor.kt:74)
	at org.spekframework.spek2.junit.SpekTestDescriptor.removeFromHierarchy(SpekTestDescriptor.kt:11)
	at org.junit.platform.engine.TestDescriptor.prune(TestDescriptor.java:219)
	at org.junit.platform.engine.TestDescriptor.accept(TestDescriptor.java:239)
	at org.junit.platform.engine.TestDescriptor.lambda$accept$0(TestDescriptor.java:241)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.junit.platform.engine.TestDescriptor.accept(TestDescriptor.java:241)
	at org.junit.platform.launcher.core.Root.lambda$acceptInAllTestEngines$1(Root.java:79)
	at java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:608)
	at org.junit.platform.launcher.core.Root.acceptInAllTestEngines(Root.java:79)
	at org.junit.platform.launcher.core.Root.prune(Root.java:71)
	at org.junit.platform.launcher.core.DefaultLauncher.discoverRoot(DefaultLauncher.java:121)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:92)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$100(JUnitPlatformTestClassProcessor.java:77)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:73)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	... 25 more
```

It took me awhile to figure out what's going on because there's nothing in the stack trace that hints at a file or line within a file. I was able to narrow it down after finding the offending commit, commenting all tests out, and then uncommenting them all back in, one by one. Kind of a pain :) It would be great if empty `xdescribe` test blocks were passing or ignored by default, instead of causing the whole suite to fail. That said, the workaround is simple: don't add empty `xdescribe` blocks!

--- a/integration-test/common/src/test/kotlin/org/spekframework/spek2/integration/EmptyGroupTest.kt
+++ b/integration-test/common/src/test/kotlin/org/spekframework/spek2/integration/EmptyGroupTest.kt
@@ -1,0 +1,11 @@
+package org.spekframework.spek2.integration
+
+import org.spekframework.spek2.Spek
+
+object EmptyGroupTest: Spek({
+    group("empty group is valid") {
+        // should not cause any failures
+    }
+
+    test("there should be at least one test in the tree") {}
+})

--- a/spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestDescriptor.kt
+++ b/spek-runner/junit5/src/main/kotlin/org/spekframework/spek2/junit/SpekTestDescriptor.kt
@@ -5,7 +5,10 @@ import org.junit.platform.engine.TestSource
 import org.junit.platform.engine.TestTag
 import org.junit.platform.engine.UniqueId
 import org.junit.platform.engine.support.descriptor.ClassSource
-import org.spekframework.spek2.runtime.scope.*
+import org.spekframework.spek2.runtime.scope.GroupScopeImpl
+import org.spekframework.spek2.runtime.scope.ScopeImpl
+import org.spekframework.spek2.runtime.scope.ScopeType
+import org.spekframework.spek2.runtime.scope.TestScopeImpl
 import java.util.*
 
 class SpekTestDescriptor internal constructor(
@@ -71,8 +74,19 @@ class SpekTestDescriptor internal constructor(
 
     override fun getTags(): MutableSet<TestTag> = mutableSetOf()
 
-    override fun removeFromHierarchy() = throw UnsupportedOperationException()
-    override fun removeChild(descriptor: TestDescriptor) = throw UnsupportedOperationException()
+    override fun removeFromHierarchy() {
+        parent.ifPresent { parent ->
+            parent.removeChild(this)
+        }
+    }
+
+    override fun removeChild(descriptor: TestDescriptor) {
+        if (scope is GroupScopeImpl) {
+            childDescriptors.remove(descriptor)
+            scope.removeChild((descriptor as SpekTestDescriptor).scope)
+        }
+    }
+
     override fun findByUniqueId(uniqueId: UniqueId): Optional<out TestDescriptor> =
         throw UnsupportedOperationException()
 }

--- a/spek-runtime/common/src/main/kotlin/org/spekframework/spek2/runtime/scope/Scopes.kt
+++ b/spek-runtime/common/src/main/kotlin/org/spekframework/spek2/runtime/scope/Scopes.kt
@@ -48,6 +48,10 @@ open class GroupScopeImpl(
         children.add(child)
     }
 
+    fun removeChild(child: ScopeImpl) {
+        children.remove(child)
+    }
+
     fun getChildren() = children.toList()
 
     fun filterBy(path: Path) {

